src/test/net_tests.cpp

```diff
@@ -771,4 +773,129 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+    Shuffle(shuffled_candidates.begin(), shuffled_candidates.end(), random_context);
+    const Optional<NodeEvictionCandidate> evicted_node = SelectNodeToEvict(std::move(shuffled_candidates));
+    if (!evicted_node) {
+        return false;
+    }
+    return std::find(node_ids.begin(), node_ids.end(), evicted_node->id) != node_ids.end();
+}
+
+bool IsEvicted(const int number_of_nodes, std::function<void(NodeEvictionCandidate&, const size_t)> candidate_setup_fn, const std::vector<NodeId> node_ids, FastRandomContext& random_context)
+{
+    std::vector<NodeEvictionCandidate> candidates = GetRandomNodeEvictionCandidates(number_of_nodes, random_context);
+    for (size_t i = 0; i < candidates.size(); ++i) {
+        candidate_setup_fn(candidates[i], i);
+    }
+    return IsEvicted(candidates, node_ids, random_context);
+}
+
+namespace {
+// The number of eviction candidates at (or above) which we are guaranteed that
+// the current eviction logic will find a node to evict no matter what eviction
+// candidates it is given. This constant may need to be adjusted if the eviction
+// logic changes.
+const int GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE = 29;
+
+// The number of eviction candidates at (or below) which we are guaranteed that
+// the current eviction logic won't find a node to evict no matter what eviction
+// candidates it is given. This constant may need to be adjusted if the eviction
+// logic changes.
+const int GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW = 20;
+
+const int NODE_EVICTION_TEST_ROUNDS = 10;
+
+const int NODE_EVICTION_TEST_UP_TO_N_NODES = 200;
+} // namespace
+
+static_assert(GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE > GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW);
+
+BOOST_AUTO_TEST_CASE(node_eviction_test)
+{
+    FastRandomContext random_context{true};
+
+    for (int i = 0; i < NODE_EVICTION_TEST_ROUNDS; ++i) {
+        for (int number_of_nodes = 0; number_of_nodes < NODE_EVICTION_TEST_UP_TO_N_NODES; ++number_of_nodes) {
+            // Verify correctness of GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE
+            // and GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW.
+            if (number_of_nodes <= GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW) {
+                BOOST_CHECK(!SelectNodeToEvict(GetRandomNodeEvictionCandidates(number_of_nodes, random_context)));
+            }
+            if (number_of_nodes >= GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE) {
+                BOOST_CHECK(SelectNodeToEvict(GetRandomNodeEvictionCandidates(number_of_nodes, random_context)));
+            } else {
+                continue;
+            }
+
+            // Four nodes with the highest keyed netgroup values should be
+            // protected from eviction.
+            BOOST_CHECK(!IsEvicted(
+                number_of_nodes, [number_of_nodes](NodeEvictionCandidate& candidate, const size_t idx) {
+                    candidate.id = idx;
```

Comment:
I think if you just constructed the `NodeEvictionCandidate`s with sequential node id numbers, you wouldn't need to do this here.